olive/model/handler/qnn.py

```diff
@@ -0,0 +1,100 @@
+
+from olive.common.config_utils import validate_config
+from olive.constants import Framework, ModelFileFormat
+from olive.hardware.accelerator import Device
+from olive.model.config import IoConfig
+from olive.model.config.registry import model_handler_registry
+from olive.model.handler.base import OliveModelHandler
+from olive.platform_sdk.qualcomm.qnn.qnn import QNNInferenceSession, QNNSessionOptions
+
+logger = logging.getLogger(__name__)
+
+
+@model_handler_registry("QNNModel")
+class QNNModelHandler(OliveModelHandler):
+    json_config_keys: Tuple[str, ...] = ("io_config", "model_file_format")
+
+    def __init__(
+        self,
+        model_path: str,
+        model_attributes: Optional[Dict[str, Any]] = None,
+        io_config: Union[Dict[str, Any], IoConfig, str, Callable] = None,
+        model_file_format: ModelFileFormat = ModelFileFormat.QNN_CPP,
+    ):
+        super().__init__(
+            framework=Framework.QNN,
+            model_file_format=model_file_format,
+            model_path=model_path,
+            model_attributes=model_attributes,
+        )
+        # io config for conversion to onnx
+        self.io_config = (
+            validate_config(io_config, IoConfig).dict() if isinstance(io_config, (IoConfig, dict)) else io_config
+        )
+
+    @property
+    def model_path(self):
+        model_path = super().model_path
+        if self.model_file_format == ModelFileFormat.QNN_CPP:
+            logger.debug("QNNModelHandler: model_path is the cpp file for QNN_CPP model format.")
+        elif self.model_file_format == ModelFileFormat.QNN_LIB:
+            # self.model_path is the folder containing the lib file, the structure is like:
+            # - self.model_path
+            #   - aarch64-android
+            #     - libmodel.so
+            #   - x86_64-linux-clang
+            #     - libmodel.so
+            model_attributes = self.model_attributes or {}
+            model_lib_suffix = None
+            lib_targets = model_attributes.get("lib_targets")
+            if lib_targets is None:
+                logger.debug(
+                    "QNNModelHandler: lib_targets is not provided, using default lib_targets x86_64-linux-clang"
+                )
+                if platform.system() == "Linux":
+                    lib_targets = "x86_64-linux-clang"
+                    model_lib_suffix = ".so"
+                elif platform.system() == "Windows":
+                    # might be different for arm devices
+                    lib_targets = "x64"
+                    model_lib_suffix = ".dll"
+            model_folder = Path(model_path) / lib_targets
+            model_paths = list(model_folder.glob(f"*{model_lib_suffix}"))
+            if not model_paths:
+                raise FileNotFoundError(f"No model file found in {model_folder}")
+            elif len(model_paths) > 1:
+                raise RuntimeError(f"Multiple model files found in {model_folder}: {model_paths}")
```

Comment:
should the optimization pass since the RuntimeError will be ignored? or need we raise assertionerror so the optimization will fail?

Reply:
Make sense. Updated both case with assertion.
![image](https://github.com/microsoft/Olive/assets/13343117/e8a492ff-3565-4e50-ba06-29f8cf823072)